lib/sparkql/function_resolver.rb

```diff
@@ -1,804 +1,855 @@
+      # Wrap this string in quotes, as we effectively translate
+      #   City Eq endswith('far')
+      # ...to...
+      #    City Eq regex('far$')
+      #
+      # The string passed in will merely be "far", rather than
+      # the string literal "'far'".
+      string = Regexp.escape(string)
+      new_value = "#{string}$"
+
+      {
+        function_name: 'regex',
+        function_parameters: [new_value, ''],
+        type: :character,
+        value: new_value
+      }
     end
 
-    new_string = character[first_index..second_index].to_s
+    def contains(string)
+      # Wrap this string in quotes, as we effectively translate
+      #   City Eq contains('far')
+      # ...to...
+      #    City Eq regex('far')
+      #
+      # The string passed in will merely be "far", rather than
+      # the string literal "'far'".
+      string = Regexp.escape(string)
+      new_value = string.to_s
 
-    {
-      :type => :character,
-      :value => new_string
-    }
-  end
+      {
+        function_name: 'regex',
+        function_parameters: [new_value, ''],
+        type: :character,
+        value: new_value
+      }
+    end
 
-  def substring_index_error?(second_index)
-    if second_index.to_i < 0
-      @errors << Sparkql::ParserError.new(:token => second_index,
-        :message => "Function call 'substring' may not have a negative integer for its second parameter",
-        :status => :fatal)
-      true
+    # Offset the current timestamp by a number of seconds
+    def seconds(num)
+      t = current_time + num
+      {
+        type: :datetime,
+        value: t.iso8601
+      }
     end
-    false
-  end
 
-  def tolower(args)
-    {
-      :type => :character,
-      :value => "tolower"
-    }
-  end
+    # Offset the current timestamp by a number of minutes
+    def minutes(num)
+      t = current_time + num * SECONDS_IN_MINUTE
+      {
+        type: :datetime,
+        value: t.iso8601
+      }
+    end
 
-  def tolower_character(string)
-    {
-      :type => :character,
-      :value => "'#{string.to_s.downcase}'"
-    }
-  end
+    # Offset the current timestamp by a number of hours
+    def hours(num)
+      t = current_time + num * SECONDS_IN_HOUR
+      {
+        type: :datetime,
+        value: t.iso8601
+      }
+    end
 
+    # Offset the current timestamp by a number of days
+    def days(number_of_days)
+      # date calculated as the offset from midnight tommorrow. Zero will provide values for all times
+      # today.
+      d = current_date + number_of_days
+      {
+        type: :date,
+        value: d.strftime(STRFTIME_DATE_FORMAT)
+      }
+    end
 
-  def toupper_character(string)
-    {
-      :type => :character,
-      :value => "'#{string.to_s.upcase}'"
-    }
-  end
+    def weekdays(number_of_days)
+      today = current_date
+      weekend_start = today.saturday? || today.sunday?
+      direction = number_of_days.positive? ? 1 : -1
+      weeks = (number_of_days / 5.0).to_i
```

Comment:
Shout out to Mr. Lobdell who helped me realize  `-4/5 == -1`. This is the workaround I came up with to drop the remainder.